src/Grid.ts

```diff
@@ -0,0 +1,110 @@
+import { Subscription } from '@dojo/shim/Observable';
+import { v, w } from '@dojo/widget-core/d';
+import { DNode, PropertyChangeRecord } from '@dojo/widget-core/interfaces';
+import { RegistryMixin }  from '@dojo/widget-core/mixins/Registry';
+import { theme, ThemeableMixin, ThemeableProperties } from '@dojo/widget-core/mixins/Themeable';
+import WidgetBase, { diffProperty } from '@dojo/widget-core/WidgetBase';
+import WidgetRegistry from '@dojo/widget-core/WidgetRegistry';
+import DataProviderBase, { Options } from './bases/DataProviderBase';
+import Body from './Body';
+import Cell from './Cell';
+import Header from './Header';
+import HeaderCell from './HeaderCell';
+import { DataProperties, HasColumns } from './interfaces';
+import Row from './Row';
+
+import * as css from './styles/grid.m.css';
+
+export const GridBase = ThemeableMixin(RegistryMixin(WidgetBase));
+
+/**
+ * @type GridProperties
+ *
+ * Properties that can be set on a Grid
+ *
+ * @property columns		Column definitions
+ * @property dataProvider	An observable object that responds to events and returns {@link DataProperties}
+ */
+export interface GridProperties extends ThemeableProperties, HasColumns {
+	registry?: WidgetRegistry;
+	dataProvider: DataProviderBase<any, Options>;
+}
+
+const gridRegistry = new WidgetRegistry();
+gridRegistry.define('header', Header);
+gridRegistry.define('header-cell', HeaderCell);
+gridRegistry.define('body', Body);
+gridRegistry.define('row', Row);
+gridRegistry.define('cell', Cell);
+
+@theme(css)
+class Grid extends GridBase<GridProperties> {
+	private _data: DataProperties<any>;
```

Comment:
I could probably use `object` now instead of `any` - but the grid doesn't deal with the underlying data directly other than reading a field from that object. The reason I was using a generic is that the user creating the `DataProvider` will have access to that type when observing.